File: backend/app/schemas/email.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Optional, List
from enum import Enum


# --- SQLAlchemy側のEnumと一致するEnumを定義 ---
class EmailCategory(str, Enum):
    unknown = "unknown"
    job = "job"
    engineer = "engineer"


# --- GeminiのAPIのレスポンス型 ---
class EmailResponseGemini(BaseModel):
    category: EmailCategory
    skills: List[str]
    price_min: int
    price_max: int


# --- APIレスポンスモデル 添付ファイル ---
class EmailAttachmentsResponse(BaseModel):
    id: int
    email_id: int
    file_name: str
    file_path: str
    mime_type: str
    size: int
    model_config = ConfigDict(from_attributes=True)


# --- APIレスポンスモデル Eメール ---
class EmailResponse(BaseModel):
    id: int
    uid: str
    from_address: str
    subject: Optional[str]
    received_at: datetime
    body: Optional[str]
    category: EmailCategory
    skills: Optional[List[str]] = None  # JSON型はDictで表現
    price_min: Optional[int]
    price_max: Optional[int]
    ai_confidence: Optional[float]
    attachments: List[EmailAttachmentsResponse] = []
    model_config = ConfigDict(from_attributes=True)


# --- router関数の引数型 ---
class SearchRequest(BaseModel):
    query: str
    category: str


class MatchRequest(BaseModel):
    category: str
    skills: List[str]


class EmailSyncStatusResponse(BaseModel):
    last_updated_at: datetime


class DownloadFileRequest(BaseModel):
    uid: str
    file_name: str
    mime_type: str

File: frontend/src/types/email.ts
export type MailCategory = "engineer" | "job";
import type { Dispatch, SetStateAction } from "react";

export type Email = {
  id: number;
  uid: string;
  from_address: string;
  subject: string;
  received_at: string;
  category: MailCategory;
  body?: string;
  skills?: string[];
  // price_min?: number;
  // price_max?: number;
  // ai_confidence?: number;
  attachments: Attachment[];
};

export type Attachment = {
  id: number;
  email_id: number;
  file_name: string;
  file_path: string;
  mime_type: string;
  size: number;
};

export type EmailDetailProps = {
  email: Email;
  onBack: () => void;
  onDownloadFile: (email_id: string, attachment: Attachment) => void;
};

export type EmailListProps = {
  title: string;
  emails: Email[] | null;
  onEmailClick: (email: Email) => void;
  scrollPosition: number;
  onScrollChange: (pos: number) => void;
  currentPage: number;
  onPageChange: (p: number) => void;
};

export type SearchSectionProps = {
  leftSkills: string[];
  rightSkills: string[];
  selectedSkills: string[];
  onToggleSkill: (skill: string) => void;
  onMatchEmails: () => Promise<void>;
  onResetMatch: () => void;
  newSkill: string;
  setNewSkill: (skill: string) => void;
  isAdding: boolean;
  setIsAdding: Dispatch<SetStateAction<boolean>>;
  matchedEmails: Email[] | null;
  activeTab: MailCategory;
  setQuery: (query: string) => void;
  onSearch: () => void;
  onResetSearch: () => void;
};

export type SkillsTagSectionProps = {
  skills: string[];
  selectedSkills: string[];
  newSkill?: string;
  isAdding?: boolean;
  editable?: boolean; // ← 左側のみ true（追加機能あり）
  onToggleSkill?: (skill: string) => void;
  onAddSkill?: () => void;
  onToggleIsAdding?: () => void;
  onChangeSkillInput?: (value: string) => void;
  title: string;
};

export type SearchedEmails = {
  engineer?: Email[];
  job?: Email[];
};

export type EmailSyncStatus = {
  last_updated_at: string | null;
};
